File: sdk/python/eventrelay_signature.py
"""Dependency-free EventRelay HMAC verifier."""
import hashlib
import hmac
import time


def verify(secret: str, timestamp_ms: str, event_id: str, payload: bytes,
           signature_header: str, tolerance_seconds: int = 300) -> bool:
    try:
        timestamp = int(timestamp_ms)
        if abs(int(time.time() * 1000) - timestamp) > tolerance_seconds * 1000:
            return False
        supplied = dict(part.split("=", 1) for part in signature_header.split(","))
        signed = timestamp_ms.encode() + b"." + event_id.encode() + b"." + payload
        expected = hmac.new(secret.encode(), signed, hashlib.sha256).hexdigest()
        return hmac.compare_digest(expected, supplied.get("v1", ""))
    except (ValueError, KeyError):
        return False

File: sdk/javascript/eventrelay-signature.js
import { createHmac, timingSafeEqual } from 'node:crypto';

export function verifyEventRelaySignature({
  secret, timestampMs, eventId, rawBody, signatureHeader, toleranceSeconds = 300,
}) {
  const timestamp = Number(timestampMs);
  if (!Number.isFinite(timestamp)
      || Math.abs(Date.now() - timestamp) > toleranceSeconds * 1000) return false;
  const parts = Object.fromEntries(signatureHeader.split(',').map((part) => part.split('=', 2)));
  const expected = createHmac('sha256', secret)
    .update(`${timestampMs}.${eventId}.`).update(rawBody).digest('hex');
  const supplied = parts.v1 || '';
  if (expected.length !== supplied.length) return false;
  return timingSafeEqual(Buffer.from(expected), Buffer.from(supplied));
}
